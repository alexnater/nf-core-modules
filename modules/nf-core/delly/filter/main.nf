process DELLY_FILTER {
    tag "$meta.id"
    label 'process_single'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/delly:1.2.6--hb7e2ac5_0':
        'biocontainers/delly:1.2.6--hb7e2ac5_0' }"

    input:
    tuple val(meta), path(bcf), path(csi)
    val(mode)
    path(samples)

    output:
    tuple val(meta), path("${prefix}.bcf")    , emit: bcf
    tuple val(meta), path("${prefix}.bcf.csi"), emit: csi
    path "versions.yml"                       , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    if ("$bcf" == "${prefix}.bcf") error "Input and output names are the same, set prefix in module configuration to disambiguate!"
    def samples_arg = ''
    def sample_list = ''
    if (mode == 'somatic') {
        if (!samples && meta.control_samples && meta.tumor_samples) {
            sample_list = meta.control_samples.collect {"${it}\tcontrol"}.join('\n')
            sample_list += '\n' + meta.tumor_samples.collect {"${it}\ttumor"}.join('\n')
        }
        else {
            error "For somatic mode, a list of sample assignments must be provided."
        }
        samples_arg = '-s samples.tsv'
    } else if (mode != 'germline') {
        error "Mode must be one of 'germline' or 'somatic'."
    }

    """
    if [ "$samples_arg" ]; then echo "$sample_list" > samples.tsv; fi
    
    delly \\
        filter \\
        $args \\
        -f $mode \\
        -o ${prefix}.bcf \\
        $samples_arg \\
        $bcf
    
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        delly: \$( echo \$(delly --version 2>&1) | sed 's/^.*Delly version: v//; s/ using.*\$//')
    END_VERSIONS
    """

    stub:
    prefix = task.ext.prefix ?: "${meta.id}"

    """
    touch ${prefix}.bcf
    touch ${prefix}.bcf.csi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        delly: \$( echo \$(delly --version 2>&1) | sed 's/^.*Delly version: v//; s/ using.*\$//')
    END_VERSIONS
    """
}