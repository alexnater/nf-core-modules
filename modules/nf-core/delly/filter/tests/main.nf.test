nextflow_process {

    name "Test Process DELLY_FILTER"
    script "../main.nf"
    process "DELLY_FILTER"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "delly"
    tag "delly/call"
    tag "delly/filter"

    test("germline - bcf") {
        setup {
            run("DELLY_CALL") {
                script "../../call/main.nf"
                    process {
                        """
                        input[0] = [
                            [ id:'control' ], // meta map
                            file(params.test_data['homo_sapiens']['illumina']['test_paired_end_recalibrated_sorted_bam'], checkIfExists: true),
                            file(params.test_data['homo_sapiens']['illumina']['test_paired_end_recalibrated_sorted_bam_bai'], checkIfExists: true),
                            [],
                            [],
                            []
                            ]
                        input[1] = [
                            [ id:'human' ],
                            file(params.test_data['homo_sapiens']['genome']['genome_21_fasta'], checkIfExists: true),
                            file(params.test_data['homo_sapiens']['genome']['genome_21_fasta_fai'], checkIfExists: true)
                            ]
                        input[2] = [
                            [:],
                            [],
                            []
                            ]
                        """
                    }
                }
            }

        when {
            process {
                """
                input[0] = DELLY_CALL.out.bcf.join(DELLY_CALL.out.csi)
                input[1] = "germline"
                input[2] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.versions }
            )
        }
    }

    test("germline - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    [],
                    []
                    ]
                input[1] = "germline"
                input[2] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.versions }
            )
        }

    }

    test("somatic - bcf") {
        setup {
            run("DELLY_CALL") {
                script "../../call/main.nf"
                    process {
                        """
                        input[0] = [
                            [ id:'tumor_control' ], // meta map
                            file(params.test_data['homo_sapiens']['illumina']['test2_paired_end_recalibrated_sorted_bam'], checkIfExists: true),
                            file(params.test_data['homo_sapiens']['illumina']['test2_paired_end_recalibrated_sorted_bam_bai'], checkIfExists: true),
                            file(params.test_data['homo_sapiens']['illumina']['test_paired_end_recalibrated_sorted_bam'], checkIfExists: true),
                            file(params.test_data['homo_sapiens']['illumina']['test_paired_end_recalibrated_sorted_bam_bai'], checkIfExists: true),
                            []
                            ]
                        input[1] = [
                            [ id:'human' ],
                            file(params.test_data['homo_sapiens']['genome']['genome_21_fasta'], checkIfExists: true),
                            file(params.test_data['homo_sapiens']['genome']['genome_21_fasta_fai'], checkIfExists: true)
                            ]
                        input[2] = [
                            [:],
                            [],
                            []
                            ]
                        """
                    }
                }
            }
        when {
            process {
                """
                ch_input = DELLY_CALL.out.bcf
                            .join(DELLY_CALL.out.csi)
                            .map { meta, bcf, csi ->
                                def new_meta = [
                                    id: meta.id,
                                    tumor_samples: ['tumour'],
                                    control_samples: ['normal']
                                  ]
                                [ new_meta, bcf, csi ]
                            }
                input[0] = ch_input
                input[1] = "somatic"
                input[2] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.versions }
            )
        }
    }
}
