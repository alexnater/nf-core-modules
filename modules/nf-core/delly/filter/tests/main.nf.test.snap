{
    "somatic - bcf": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "tumor_control",
                            "tumor_samples": [
                                "tumour"
                            ],
                            "control_samples": [
                                "normal"
                            ]
                        },
                        "tumor_control.filtered.bcf:md5,2409539492ce9815ad05363e5a80ca62"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "tumor_control",
                            "tumor_samples": [
                                "tumour"
                            ],
                            "control_samples": [
                                "normal"
                            ]
                        },
                        "tumor_control.filtered.bcf.csi:md5,66545c77fc791a02dadc7fc252d04635"
                    ]
                ],
                "2": [
                    "versions.yml:md5,f67ef5fd2fa49a82ccea58f0d3d5ad0f"
                ],
                "bcf": [
                    [
                        {
                            "id": "tumor_control",
                            "tumor_samples": [
                                "tumour"
                            ],
                            "control_samples": [
                                "normal"
                            ]
                        },
                        "tumor_control.filtered.bcf:md5,2409539492ce9815ad05363e5a80ca62"
                    ]
                ],
                "csi": [
                    [
                        {
                            "id": "tumor_control",
                            "tumor_samples": [
                                "tumour"
                            ],
                            "control_samples": [
                                "normal"
                            ]
                        },
                        "tumor_control.filtered.bcf.csi:md5,66545c77fc791a02dadc7fc252d04635"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,f67ef5fd2fa49a82ccea58f0d3d5ad0f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-21T17:17:34.432098148"
    },
    "germline - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.filtered.bcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.filtered.bcf.csi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,f45104c406e3750e1f09c3ead45a8829"
                ],
                "bcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.filtered.bcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "csi": [
                    [
                        {
                            "id": "test"
                        },
                        "test.filtered.bcf.csi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,f45104c406e3750e1f09c3ead45a8829"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-21T17:17:18.241522976"
    },
    "germline - bcf": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "control"
                        },
                        "control.filtered.bcf:md5,5d8edd5f86799b2c16b6ce76ef16ff9c"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "control"
                        },
                        "control.filtered.bcf.csi:md5,66545c77fc791a02dadc7fc252d04635"
                    ]
                ],
                "2": [
                    "versions.yml:md5,f45104c406e3750e1f09c3ead45a8829"
                ],
                "bcf": [
                    [
                        {
                            "id": "control"
                        },
                        "control.filtered.bcf:md5,5d8edd5f86799b2c16b6ce76ef16ff9c"
                    ]
                ],
                "csi": [
                    [
                        {
                            "id": "control"
                        },
                        "control.filtered.bcf.csi:md5,66545c77fc791a02dadc7fc252d04635"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,f45104c406e3750e1f09c3ead45a8829"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-21T17:17:06.385722493"
    }
}